pyOCD/coresight/ap.py

```diff
@@ -0,0 +1,347 @@
+    def read32(self, addr, now=True):
+        return self.readMemory(addr, 32, now)
+
+    ## @brief Shorthand to read a 16-bit halfword.
+    def read16(self, addr, now=True):
+        return self.readMemory(addr, 16, now)
+
+    ## @brief Shorthand to read a byte.
+    def read8(self, addr, now=True):
+        return self.readMemory(addr, 8, now)
+
+    ## @brief Read a block of unaligned bytes in memory.
+    # @return an array of byte values
+    def readBlockMemoryUnaligned8(self, addr, size):
+        res = []
+
+        # try to read 8bits data
+        if (size > 0) and (addr & 0x01):
+            mem = self.readMemory(addr, 8)
+            res.append(mem)
+            size -= 1
+            addr += 1
+
+        # try to read 16bits data
+        if (size > 1) and (addr & 0x02):
+            mem = self.readMemory(addr, 16)
+            res.append(mem & 0xff)
+            res.append((mem >> 8) & 0xff)
+            size -= 2
+            addr += 2
+
+        # try to read aligned block of 32bits
+        if (size >= 4):
+            mem = self.readBlockMemoryAligned32(addr, size/4)
+            res += conversion.u32leListToByteList(mem)
+            size -= 4*len(mem)
+            addr += 4*len(mem)
+
+        if (size > 1):
+            mem = self.readMemory(addr, 16)
+            res.append(mem & 0xff)
+            res.append((mem >> 8) & 0xff)
+            size -= 2
+            addr += 2
+
+        if (size > 0):
+            mem = self.readMemory(addr, 8)
+            res.append(mem)
+            size -= 1
+            addr += 1
+
+        return res
+
+    ## @brief Write a block of unaligned bytes in memory.
+    def writeBlockMemoryUnaligned8(self, addr, data):
+        size = len(data)
+        idx = 0
+
+        #try to write 8 bits data
+        if (size > 0) and (addr & 0x01):
+            self.writeMemory(addr, data[idx], 8)
+            size -= 1
+            addr += 1
+            idx += 1
+
+        # try to write 16 bits data
+        if (size > 1) and (addr & 0x02):
+            self.writeMemory(addr, data[idx] | (data[idx+1] << 8), 16)
+            size -= 2
+            addr += 2
+            idx += 2
+
+        # write aligned block of 32 bits
+        if (size >= 4):
+            data32 = conversion.byteListToU32leList(data[idx:idx + (size & ~0x03)])
+            self.writeBlockMemoryAligned32(addr, data32)
+            addr += size & ~0x03
+            idx += size & ~0x03
+            size -= size & ~0x03
+
+        # try to write 16 bits data
+        if (size > 1):
+            self.writeMemory(addr, data[idx] | (data[idx+1] << 8), 16)
+            size -= 2
+            addr += 2
+            idx += 2
+
+        #try to write 8 bits data
+        if (size > 0):
+            self.writeMemory(addr, data[idx], 8)
+            size -= 1
+            addr += 1
+            idx += 1
+
+        return
+
+    ## @brief Write a block of aligned words in memory.
+    def writeBlockMemoryAligned32(self, addr, data):
```

Comment:
I'm not sure if there is much of a need to write blocks of 32 bytes values.  It might be good to make this private/remove it from the public API.  What do you think?